### ISSUE TITLE
DNS may not be available during kubedns-relation causing failures

### ISSUE THREAD

**Issue by marcoceppi** [juju-solutions/bundle-canonical-kubernetes#122]
In the following deployment:

```
2016-11-04 09:18:21 INFO kube-dns-relation-changed switched to context "juju-context".
2016-11-04 09:18:21 INFO juju-log kube-dns:10: Writing file /etc/default/kube-default root:root 444
2016-11-04 09:18:21 INFO juju-log kube-dns:10: Writing file /etc/default/kubelet root:root 444
2016-11-04 09:18:21 INFO juju-log kube-dns:10: Writing file /etc/default/kube-proxy root:root 444
2016-11-04 09:18:21 INFO juju-log kube-dns:10: Writing file /lib/systemd/system/kube-proxy.service root:root 444
2016-11-04 09:18:21 INFO juju-log kube-dns:10: Writing file /lib/systemd/system/kubelet.service root:root 444
2016-11-04 09:18:21 INFO kube-dns-relation-changed Created symlink from /etc/systemd/system/multi-user.target.wants/kubelet.service to /lib/systemd/system/kubelet.service.
2016-11-04 09:18:21 INFO kube-dns-relation-changed Created symlink from /etc/systemd/system/multi-user.target.wants/kube-proxy.service to /lib/systemd/system/kube-proxy.service.
2016-11-04 09:18:22 INFO juju-log kube-dns:10: Restarting kubelet, and kube-proxy.
2016-11-04 09:18:22 INFO juju-log kube-dns:10: Invoking reactive handler: reactive/kubernetes_worker.py:84:set_app_version
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Invoking reactive handler: reactive/kubernetes_worker.py:175:render_and_launch_ingress
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Making dir /etc/kubernetes/addons root:root 755
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Writing file /etc/kubernetes/addons/default-http-backend.yaml root:root 444
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Creating the default http backend.
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Executing ['kubectl', '--kubeconfig=/srv/kubernetes/config', 'get', '-f', '/etc/kubernetes/addons/default-http-backend.yaml']
2016-11-04 09:18:23 INFO kube-dns-relation-changed Error from server: an error on the server ("<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>") has prevented the request from succeeding
2016-11-04 09:18:23 INFO juju-log kube-dns:10: Executing ['kubectl', '--kubeconfig=/srv/kubernetes/config', 'create', '-f', '/etc/kubernetes/addons/default-http-backend.yaml']
2016-11-04 09:18:24 INFO kube-dns-relation-changed Error from server: an error on the server ("<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>") has prevented the request from succeeding
2016-11-04 09:18:24 INFO juju-log kube-dns:10: Writing file /etc/kubernetes/addons/ingress-replication-controller.yaml root:root 444
2016-11-04 09:18:24 INFO juju-log kube-dns:10: Executing ['kubectl', '--kubeconfig=/srv/kubernetes/config', 'get', '-f', '/etc/kubernetes/addons/ingress-replication-controller.yaml']
2016-11-04 09:18:24 INFO kube-dns-relation-changed Error from server: an error on the server ("<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>") has prevented the request from succeeding
2016-11-04 09:18:24 INFO juju-log kube-dns:10: Executing ['kubectl', '--kubeconfig=/srv/kubernetes/config', 'create', '-f', '/etc/kubernetes/addons/ingress-replication-controller.yaml']
2016-11-04 09:18:24 INFO kube-dns-relation-changed Error from server: an error on the server ("<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>") has prevented the request from succeeding
2016-11-04 09:18:24 INFO juju-log kube-dns:10: Creating the ingress replication controller.
2016-11-04 09:18:25 INFO juju-log kube-dns:10: Invoking reactive handler: reactive/kubernetes_worker.py:196:scale_ingress_controller
2016-11-04 09:18:25 INFO juju-log kube-dns:10: Executing ['kubectl', '--kubeconfig=/srv/kubernetes/config', 'get', 'nodes', '-o', 'name']
2016-11-04 09:18:25 INFO kube-dns-relation-changed Error from server: an error on the server ("<html>\r\n<head><title>502 Bad Gateway</title></head>\r\n<body bgcolor=\"white\">\r\n<center><h1>502 Bad Gateway</h1></center>\r\n<hr><center>nginx/1.10.0 (Ubuntu)</center>\r\n</body>\r\n</html>") has prevented the request from succeeding
2016-11-04 09:18:25 INFO kube-dns-relation-changed Traceback (most recent call last):
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/var/lib/juju/agents/unit-kubernetes-worker-0/charm/hooks/kube-dns-relation-changed", line 19, in <module>
2016-11-04 09:18:25 INFO kube-dns-relation-changed     main()
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/__init__.py", line 78, in main
2016-11-04 09:18:25 INFO kube-dns-relation-changed     bus.dispatch()
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 434, in dispatch
2016-11-04 09:18:25 INFO kube-dns-relation-changed     _invoke(other_handlers)
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 417, in _invoke
2016-11-04 09:18:25 INFO kube-dns-relation-changed     handler.invoke()
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/local/lib/python3.5/dist-packages/charms/reactive/bus.py", line 291, in invoke
2016-11-04 09:18:25 INFO kube-dns-relation-changed     self._action(*args)
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/var/lib/juju/agents/unit-kubernetes-worker-0/charm/reactive/kubernetes_worker.py", line 200, in scale_ingress_controller
2016-11-04 09:18:25 INFO kube-dns-relation-changed     output = kubectl('get', 'nodes', '-o', 'name')
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/var/lib/juju/agents/unit-kubernetes-worker-0/charm/reactive/kubernetes_worker.py", line 349, in kubectl
2016-11-04 09:18:25 INFO kube-dns-relation-changed     return check_output(command)
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/lib/python3.5/subprocess.py", line 626, in check_output
2016-11-04 09:18:25 INFO kube-dns-relation-changed     **kwargs).stdout
2016-11-04 09:18:25 INFO kube-dns-relation-changed   File "/usr/lib/python3.5/subprocess.py", line 708, in run
2016-11-04 09:18:25 INFO kube-dns-relation-changed     output=stdout, stderr=stderr)
2016-11-04 09:18:25 INFO kube-dns-relation-changed subprocess.CalledProcessError: Command '['kubectl', '--kubeconfig=/srv/kubernetes/config', 'get', 'nodes', '-o', 'name']' returned non-zero exit status 1
```

While Juju will retry automatically, we should be capturing subprocess failures and just setting state to waiting or doing a second retry/backoff.

Given we don't want to lose state, retrying during the hook context for X tries before setting state as waiting and then running again during the next hook call seems best.

**Comment by mbruzek** [comment]
I also noticed this problem.  When looking at the code we need to gate the `scale_ingress_controller` method with a "kubedns.available" state or something like that.

**Comment by chuckbutler** [comment]
The proposed fix didn't require state manipulation :)  Closed in juju-solutions/kubernetes#40
